### PR TITLE
Move most of the motd into its own window/button

### PR DESCRIPTION
**Pull request by Pentarctagon**
Also prevents the motd from being displayed to the user every time they log out then log back in.

The Windows and macOS jobs will of course fail, as this adds a new source file.

I'm not sure about the boost unit tests, though.  I added it to the omitted list of widgets since there's quite a few there already and I don't know what would need to be done to add it, but ideally it would be included in the tests.

Beyond that, the relevant information is in the commit message.

---

The window and buttons on screens larger than 800x600:

buttons-not-800x600
window-not-800x600
---

And on screens that are 800x600:

buttons-800x600
window-800x600

### CODE REVIEW

File: src/gui/widgets/chatbox.cpp [hunk 1]
```diff
@@ -754,6 +754,15 @@ void chatbox::process_message(const ::config& data, bool whisper /*= false*/)
 			room = "lobby";
 		}
 
+		if(log_ != nullptr && data["type"].str() == "motd") {
+			if(log_->at("lobby").received_motd) {
```
Comment (soliton-) [inline src/gui/widgets/chatbox.cpp:758]: This should check that it's actually the same message. You could make `received_motd` into a string.

**Comment by soliton-** [comment]
What's the advantage of putting server_info and announcments into the DB instead of the config file?

**Comment by Pentarctagon** [comment]
It was partially because then updated links, etc would be available without needing to refresh the config, and then also partially because I have access to update the database myself.

**Comment by gfgtdf** [comment]
So when is the window shown then? When you log in or only when you press the button?

**Comment by Pentarctagon** [comment]
Only when the button is pressed.

**Comment by gfgtdf** [comment]
> Only when the button is pressed.

I don't think this is optimal the user should be notified when the text changed. So what i suggest is to store the last shown string in the client sided preferences file, and when the sended message is sidderent from the stored one either just show the message directly or make the Information button blink red agressiveley.

**Comment by Pentarctagon** [comment]
Forcing a popup seems not user friendly, and neither does the blinking either (can GUI2 even do blinking buttons?).

**Comment by gfgtdf** [comment]
> Forcing a popup seems not user friendly, and neither does the blinking either

Well its not like this information changes often (i think? actually not sure), so it seems totally acceptable to me

> can GUI2 even do blinking buttons

Well we just have to implement it.

File: src/gui/widgets/chatbox.cpp (hunk 1 above)
Comment (Pentarctagon) [reply src/gui/widgets/chatbox.cpp]: Updated.

**Comment by Pentarctagon** [comment]
The frequency is not really relevant to me.  I don't like popups or things blinking at me that I need to click to make stop blinking - they're annoying.

**Comment by gfgtdf** [comment]
> The frequency is not really relevant to me. I don't like sudden popups or things blinking at me that I need to click to make go away/stop blinking - they're annoying.

Well but whats the point of "announcement" when noone will see them because no one knows that something was announced? and well i'm not sure for what type of information you wanted to announce, but if its important it will probably outwight "being annoying"

**Comment by Pentarctagon** [comment]
It would be:
* General server information, such as a link to the CoC.
* Announcements, such as tournaments and new versions being released.

From that example, tournaments probably aren't going to be of interest to most people on the MP server, and we also don't currently tell people who login to the MP server about new versions now anyway.

**Comment by soliton-** [comment]
I do agree that this needs a bit more fleshing out on what the use cases are.

With this PR the motd is only shown once it changes (not across wesnoth restarts though) so that sounds ideal to me to show occasionally changing info like tournament announcements.

Static info seem to fit well behind the "Information" button as mentioned already.

So to me the question remains if the announcement part behind the "Information" button is that useful. Having to manually check if there is something new there does not seem convenient.

> It was partially because then updated links, etc would be available without needing to refresh the config, and then also partially because I have access to update the database myself.

I don't think those are particularly good reasons. Re-reading the config file is only done once when it changes so there is no need to make a DB request each login for rarely changing data. (Also there is no problem in giving you access to update the config file as well. :wink:)

I would prefer to add another config key for the general server information instead and like alluded to above use the motd key for announcements, which better fits the meaning of "message of the day" anyway.

**Comment by Pentarctagon** [comment]
Another reason to move this information elsewhere is that there's a text limit for the lobby chat, and the current motd exceeds that limit.  So right now it's not possible to adjust the motd with the `/query motd [<message>]` command from the lobby since some of it will get cut off.

If we ever end up having more than a couple announcement type things to put in the motd, and only the general MP info is moved to this new window, then the same thing will still happen even if right now there's just the one tournament going on.

**Comment by soliton-** [comment]
`/query motd <message>` only works in the simplest case of a one line message because you cannot add newlines that way. That would need a dialog where you can write a multline message. It's also not saved to the config file so the changes are gone on server restart.

This could be made to work better though I think we've gone fine without it so far. It's also more of an admin thing not so much a moderation thing... Maybe it should be removed really. Currently you can mostly just break stuff if you try it.

**Comment by soliton-** [comment]
I think it would make sense to get tournament announcement from the DB if we would check the Tournament forum posts for open and in progress tournaments...

**Comment by Pentarctagon** [comment]
Alright, so currently the situation is:
* Change - Have the general server info be read from the config rather than the database.
* Future Change - Remove the `<message>` part of the `/query motd <message>` command.
* Open Question - Should announcements be displayed as part of the motd or as another tab on the new window with the general server information?
* Open Question - Should the tournaments list be queried from the forum database?

**Comment by Pentarctagon** [comment]
For the announcements question, I think we could do both?  By which I mean, have the announcements be put into the lobby chat on login with the motd, but also available as a tab on this new window.  That way it would still be visible to everyone on login, but users could also get back to it without needing to scroll all the way back to the top of the lobby chat.

For getting a list of tournaments from querying the forum database, would this be done at the same time as (re)reading the config file?

**Comment by soliton-** [comment]
> For the announcements question, I think we could do both? By which I mean, have the announcements be put into the lobby chat on login with the motd, but also available as a tab on this new window. That way it would still be visible to everyone on login, but users could also get back to it without needing to scroll all the way back to the top of the lobby chat.

Sounds good to me. Would it be type announcement or motd? I think if we separate out the static info from the motd both are the same really.

> For getting a list of tournaments from querying the forum database, would this be done at the same time as (re)reading the config file?

That would make sense. And it could also be updated automatically periodically, like every hour or something. Maybe have a setting for the time between updates.

**Comment by Pentarctagon** [comment]
> Sounds good to me. Would it be type announcement or motd? I think if we separate out the static info from the motd both are the same really.

I think what would need to be done would be appending the announcements text plus the result of the forum database query to the motd (so they also don't get put into the lobby chat when the user logs out/logs back in) and then sent as a separate message as well so it can be put into the new window without the motd greeting text.

> That would make sense. And it could also be updated automatically periodically, like every hour or something. Maybe have a setting for the time between updates.

I'd rather not have an asynchronous timer running queries, since right now everything works on the assumption that the database connection won't be shared.

**Comment by soliton-** [comment]
Sure, I do not mean to change how we use asio. Keep it all in one thread like now.

**Comment by soliton-** [comment]
There is a timer to dump stats periodically being used here:
https://github.com/wesnoth/wesnoth/blob/c11402ea9b375c1798da06b04e2beebcc31e6249/src/server/server.cpp#L564-L568

**Comment by Pentarctagon** [comment]
What I mean though, is what would happen if the timer runs the select statement while the database connection is already in the middle of inserting a row into another table, for example?

**Comment by soliton-** [comment]
There is only one thread running. Nothing can run in parallel. If there is a DB query running the timer cannot trigger. DB queries are done blocking as well so they cannot get nested or something (that’d already be a problem then anyway).

**Comment by Pentarctagon** [comment]
As long as asio is setup to handle the timer synchronously then, I'll imitate how the stat dump does it.

**Comment by Pentarctagon** [comment]
The query to get the tournaments would be something like:
```
select
  substring(substring_index(topic_title, ']', 1), 2) as STATUS,
  concat('https://r.wesnoth.org/t', topic_id) as URL,
  substring_index(topic_title, ']', -1) as TITLE
from forum.phpbb3_topics
where forum_id = 70
  and (topic_title like '[Open]%' or topic_title like '[In Progress]%')
```
which as of right now would return:
```
+-------------+------------------------------+-------------------------------------------+
| STATUS      | URL                          | TITLE                                     |
+-------------+------------------------------+-------------------------------------------+
| In Progress | https://r.wesnoth.org/t52344 |  Double Style Tournament - 2v2 Tourney #1 |
+-------------+------------------------------+-------------------------------------------+
```
The main problem being that the current query execution logic supports returning only single values, not multiple values or multiple rows of values.  So this PR will probably be open for a bit while I figure that out.

Technically I suppose I could concat/group_concat the whole thing into a single delimited string and then split it out afterwards, but I'd rather not.

**Comment by Pentarctagon** [comment]
@soliton- @shikadiqueen @loonycyborg the most recent commit has been updated with a more complete implementation of `mysql_conn.*pp`.

**Comment by loonycyborg** [comment]
Looks nice to me.

**Comment by soliton-** [comment]
As a general question is it wise to use the MySQL connector when we're really using MariaDB? I assume it's all compatible so far but will it stay that way? Perhaps it'd be better to use something like https://github.com/viaduck/mariadbpp?

Otherwise I don't have objections to this approach if extending our own C++ wrapper is more work. There is at least reconnect support still missing to get on par with it though.

Regarding the specifics of the tournament query, I would perhaps rather do the string handling in C++ but presumably performance is not really an issue here.

**Comment by Pentarctagon** [comment]
The two problem I saw with mariadbpp were:
* It doesn't appear to be available via repository or vcpkg, so adding it as a dependency would mean anyone who wants to build wesnothd with database support enabled would need to first download and compile that themselves.
* The one open issue it has is "Documentation is unusable".

As far as reconnecting goes, I am going to re-add that as well.  The C++ connector also has the function `isValid()`, which will allow checking whether it's encountered a bad connection or a bad query compared to currently where on any failure it tries to reconnect and re-run the same query, as well as a simpler `reconnect()` function.

**Comment by Pentarctagon** [comment]
@soliton- Reconnecting is implemented.  I tested with the scenario of:
* Log in, log out
* Turn off mysql
* Try to login, failed
* Turn mysql back on
* Log in successfully
The output from that is [here](https://pastebin.com/gdginPr5).

**Comment by Pentarctagon** [comment]
It's mostly done at this point, primarily I just need to test it out more locally.  I've also split the commits to the C++ connector code is by itself in the third commit.

@soliton- What version of the connector is installable on the server?  Locally I have `libmysqlcppconn7v5`.

**Comment by soliton-** [comment]
`libmysqlcppconn7v5` version 1.1.7 is installable.

**Comment by Pentarctagon** [comment]
Alright, so shouldn't be a problem then - the changelogs for 1.1.8 and 1.1.9 don't list any changes that would be meaningful here.

**Comment by Pentarctagon** [comment]
@soliton- The reconnect code is removed now, since when I re-tried the previous test with mariadbpp:
* When the connection was down (due to the database being turned off), the `connect()` method ended up throwing its own exception rather than returning true or false, so the contents of the if-block were never reached.
* The connection seems to automatically re-establish itself.  The first query executed after turning the database back on succeeded, also without ever reaching the reconnect logic.

**Comment by Pentarctagon** [comment]
Also, so it's mentioned here rather just in code comments - this raises wesnoth's minimum required cmake version from 2.8.5 to 3.1, and building mariadbpp requires version 3.8.

**Comment by Pentarctagon** [comment]
...which is actually a problem at the moment, since the VM running wesnothd is currently using Debian stretch which has only cmake 3.7.3 [available](https://packages.debian.org/stretch/cmake).

**Pull request closed without merging**